### ISSUE TITLE
Bug?: fetch(): Uncaught TypeError: error reading a body from connection

### ISSUE THREAD

**Issue by akku1139** [denoland/deno#25992]
Version: Deno 1.46.3

I have a problem when fetching this URL.

`http(s)://www.astroarts.co.jp/article/feed.atom`

```
> await (await fetch("http://www.astroarts.co.jp/article/feed.atom")).text()
Uncaught TypeError: error reading a body from connection
    at async readableStreamCollectIntoUint8Array (ext:deno_web/06_streams.js:1066:19)
    at async consumeBody (ext:deno_fetch/22_body.js:254:9)
    at async <anonymous>:1:22
> await (await fetch("https://www.astroarts.co.jp/article/feed.atom")).text()
Uncaught TypeError: error reading a body from connection
    at async readableStreamCollectIntoUint8Array (ext:deno_web/06_streams.js:1066:19)
    at async consumeBody (ext:deno_fetch/22_body.js:254:9)
    at async <anonymous>:1:22
```

The two below can be executed without any problems.

```
await (await fetch("http://www.astroarts.co.jp/")).text()
await (await fetch("https://www.astroarts.co.jp/")).text()
```

You can see that it has been loaded halfway.

```
> const stream = (await fetch("https://www.astroarts.co.jp/article/feed.atom")).body
undefined
> for await (const chunk of stream) { console.log(chunk) }
Uint8Array(64) [
   60,  63, 120, 109, 108,  32, 118, 101, 114, 115, 105,
  111, 110,  61,  34,  49,  46,  48,  34,  32, 101, 110,
   99, 111, 100, 105, 110, 103,  61,  34,  85,  84,  70,
   45,  56,  34,  63,  62,  10,  60, 102, 101, 101, 100,
   32, 120, 109, 108, 110, 115,  61,  34, 104, 116, 116,
  112,  58,  47,  47, 119, 119, 119,  46, 119
]
Uint8Array(64) [
   51,  46, 111, 114, 103,  47,  50,  48,  48,  53,  47,
   65, 116, 111, 109,  34,  62,  10,  32,  32,  60,  97,
// -------------- many lines ----------------
Uint8Array(64) [
  136,  49,  48, 230, 151, 165, 227, 129, 148, 227, 130,
  141, 227, 128, 130,  60,  47, 115, 117, 109, 109,  97,
  114, 121,  62,  10,  32,  32,  32,  32,  60, 117, 112,
  100,  97, 116, 101, 100,  62,  50,  48,  50,  52,  45,
   48,  57,  45,  48,  51,  84,  48,  57,  58,  49,  48,
   58,  50,  51,  90,  60,  47, 117, 112, 100
]
Uint8Array(25) [
   97, 116, 101, 100,  62,  10,  32,
   32,  60,  47, 101, 110, 116, 114,
  121,  62,  10,  60,  47, 102, 101,
  101, 100,  62,  10
]
Uncaught TypeError: error reading a body from connection
    at async Object.pull (ext:deno_web/06_streams.js:938:27)
```

**Comment by akku1139** [comment]
`fetch` was successful in the Firefox 128 dev tools console and Bun.

**Comment by BlackAsLight** [comment]
I've experience this bug a few times when making use of my `@doctor/download@0.1.0` package. Only seemed to happen though when my computer went to sleep during a download.

**Comment by paullame-loreal** [comment]
Also experienced this error when working with API from the European Union regulations 
https://ec.europa.eu/growth/tools-databases/cosing/

Ended up using node specifically for this part and it worked.

Can post details if it is useful

**Comment by digitalnature** [comment]
Getting the same error if I do a relatively high number of fetch requests in parallel.

After some investigation, I found out that this problem occurs in Docker only.
If run the Deno app outside Docker, everything is fine. I even increased the number of requests and had no issue.

Could it be that Docker throttles the network badly, maybe because of the extra overhead?

**Comment by BlackAsLight** [comment]
I got this error again. It doesn't seem to be related to getting a content-length of zero as in this latest instance I can confirm it had a big content-length.

**Comment by lionel-rowe** [comment]
I regularly encounter this error when fetching data for scheduled backups with Deno (but only intermittently, the content of the data being backed up is volatile so maybe that's why).

> `fetch` was successful in the Firefox 128 dev tools console and Bun.

Works in Chrome too, so not a V8 issue. It looks like the error comes from [`hyper`](https://github.com/hyperium/hyper/blob/621d8e4d7788bfd2d62d15d40a73efae7f9a0bf0/src/error.rs#L469). There are [various issues](https://github.com/search?q=repo%3Ahyperium%2Fhyper+%22error+reading+a+body+from+connection%22&type=issues) quoting this error message, but none currently open.

Using `https://www.astroarts.co.jp/article/feed.atom` URL from OP:

```ts
import { assertEquals } from 'jsr:@std/assert'

const url = 'https://www.astroarts.co.jp/article/feed.atom'
const td = new TextDecoder(undefined, { ignoreBOM: true, fatal: true })
const CRLF = '\r\n'
const HEADER_END_DELIM = CRLF.repeat(2)

const curl = new Deno.Command('curl', { args: ['-i', url], stdout: 'piped', stderr: 'piped' })
const curlText = td.decode((await curl.spawn().output()).stdout)
const headerEndIdx = curlText.indexOf(HEADER_END_DELIM)
const headers = new Headers(curlText.slice(0, headerEndIdx).split(CRLF).slice(1).map(x => x.split(': ')))
const expectedContentLength = Number(headers.get('content-length'))
const curled = curlText.slice(headerEndIdx + HEADER_END_DELIM.length)

let fetched = ''
let bytes = 0
let err: Error | null = null

try {
    for await (const chunk of (await fetch(url)).body!) {
        bytes += chunk.byteLength
        fetched += td.decode(chunk, { stream: true })
    }
} catch (e) {
    err = e as Error
}

assertEquals(bytes, expectedContentLength) // ok
assertEquals(fetched, curled) // ok
assertEquals(err, null) // throws

```

In other words, the `content-length` header is correct, and every _contentful_ chunk is read correctly, but the last (presumably empty) chunk throws.

**Comment by paullame-loreal** [comment]
from this web page [https://ec.europa.eu/growth/tools-databases/cosing/details/34040](https://ec.europa.eu/growth/tools-databases/cosing/details/34040), I'm trying to reproduce the request made to `https://api.tech.ec.europa.eu/search-api/prod/rest/search?apiKey=285a77fd-1257-4271-8507-f0c6b2961203&text=*&pageSize=100&pageNumber=1`

## using curl ✅
```curl
curl 'https://api.tech.ec.europa.eu/search-api/prod/rest/search?apiKey=285a77fd-1257-4271-8507-f0c6b2961203&text=*&pageSize=100&pageNumber=1' \
  -H 'Accept: application/json, text/plain, */*' \
  -H 'Content-Type: multipart/form-data; boundary=----WebKitFormBoundary68AJyAOocHzsS5XL' \
  --data-raw $'------WebKitFormBoundary68AJyAOocHzsS5XL\r\nContent-Disposition: form-data; name="query"; filename="blob"\r\nContent-Type: application/json\r\n\r\n{"bool":{"must":[{"term":{"substanceId":"34040"}}]}}\r\n------WebKitFormBoundary68AJyAOocHzsS5XL--\r\n'
```

response:
```json
{"apiVersion":"2.132","terms":"*","responseTime":27,"totalResults":1,"pageNumber":1,"pageSize":100,"sort":"relevance","groupByField":null,"queryLanguage":{"language":"en","probability":0.0},"spellingSuggestion":null,"bestBets":[],"results":[{"apiVersion":"2.132","reference":"59eb65ef-7b29-4486-a311-f194224c11e6","url":"urn:test","title":null,"contentType":"text/plain","language":"en","databaseLabel":"Growth Cosmetic Ingredients and Substances","database":"GROWTH_COSING","summary":"","weight":10.608015,"groupById":null,"content":"","accessRestriction":false,"pages":null,"checksum":"E3B0C44298FC1C149AFBF4C8996FB92427AE41E4649B934CA495991B7852B855","metadata":{"nameOfCommonIngredientsGlossary":["GLYCERIN"],"refNo":[],"otherRestrictions":[],"note":[],"sccsOpinionUrls":[],"itemType":["ingredient"],"other":[],"innName":["glycerol"],"chemicalName":["Glycerine ;Glycerol (INN); Glycerol (RIFM); Glycerolum (EP); Glycyl Alcohol; 1,2,3-Propanetriol; 1,2,3-Trihydroxypropane"],"sccsOpinion":[],"refNo_letter":[],"language":["en"],"chemicalDescription":["Glycerin is the polyhydric alcohol that conforms generally to the formula"],"esST_checksum":["E3B0C44298FC1C149AFBF4C8996FB92427AE41E4649B934CA495991B7852B855"],"reference":["8B6D5A4ACA15596EB2F4AE1A4F622100"],"cosmeticRestriction":[],"ecNo":["200-289-5"],"maximumConcentration":[],"identifiedIngredient":[],"inciName":["GLYCERIN"],"officialJournalPublication":["Y"],"esST_FileName":["file.txt"],"casNo":["56-81-5"],"substanceId":["34040"],"DATASOURCE":["GROWTH_COSING"],"productTypeBodyParts":[],"annexNo":[],"phEurName":["glycerolum"],"es_ContentType":["text/plain"],"otherRegulations":[],"classificationInformation":["[]"],"functionName":["DENATURANT","HAIR CONDITIONING","HUMECTANT","ORAL CARE","SKIN PROTECTING","SOLVENT","VISCOSITY CONTROLLING","PERFUMING","FRAGRANCE","SKIN CONDITIONING - HUMECTANT"],"esDA_IngestDate":["2024-03-14T16:26:49.529+0100"],"currentVersion":["1"],"url":["urn:test"],"esST_URL":["urn:test"],"esDA_QueueDate":["2024-03-14T16:26:49.366+0100"],"refNo_digit":[],"colour":[],"wordingOfConditions":[],"inciUsaName":[],"relatedRegulations":[],"perfuming":["N"],"status":["Active"]},"enrichedMetadata":{},"children":[],"highlightedFragments":[]}],"warnings":[]}
```

## using node [v23.7.0] (fetch) ✅
```javascript
const response = await fetch("https://api.tech.ec.europa.eu/search-api/prod/rest/search?apiKey=285a77fd-1257-4271-8507-f0c6b2961203&text=*&pageSize=100&pageNumber=1", {
    "headers": {
      "accept": "application/json, text/plain, */*",
      "content-type": "multipart/form-data; boundary=----WebKitFormBoundary68AJyAOocHzsS5XL"
    },
    "body": "------WebKitFormBoundary68AJyAOocHzsS5XL\r\nContent-Disposition: form-data; name=\"query\"; filename=\"blob\"\r\nContent-Type: application/json\r\n\r\n{\"bool\":{\"must\":[{\"term\":{\"substanceId\":\"34040\"}}]}}\r\n------WebKitFormBoundary68AJyAOocHzsS5XL--\r\n",
    "method": "POST"
  });
const data = await response.json();
console.log(data);
```

```zsh
➜  Desktop node github-issue.ts
(node:17771) ExperimentalWarning: Type Stripping is an experimental feature and might change at any time
(Use `node --trace-warnings ...` to show where the warning was created)
{
  apiVersion: '2.132',
  terms: '*',
  responseTime: 11,
  totalResults: 1,
  pageNumber: 1,
  pageSize: 100,
  sort: 'relevance',
...
```

## using Deno [v2.2.2](fetch) ❌

deno 2.2.2 (stable, release, aarch64-apple-darwin)
v8 13.4.114.9-rusty
typescript 5.7.3

```zsh
➜  Desktop deno github-issue.ts 
✅ Granted net access to "api.tech.ec.europa.eu:443".
error: Uncaught (in promise) TypeError: error reading a body from connection
const data = await response.json();
             ^
    at async readableStreamCollectIntoUint8Array (ext:deno_web/06_streams.js:1095:19)
    at async consumeBody (ext:deno_fetch/22_body.js:268:9)
```

**Comment by lionel-rowe** [comment]
It looks like the causes (at least the proximate causes) of the errors are different between the `astroarts.co.jp` and `ec.europa.eu` results, as "error reading a body from connection" is a very generic error message from hyper (the internal details of the error aren't exposed by its `fmt::Display` trait).
  
* `astroarts.co.jp` has a correct `content-length` header but somehow it seems the measurement of already-read bytes vs remaining bytes gets out of sync in `hyper`, causing [this code branch](https://github.com/hyperium/hyper/blob/621d8e4d7788bfd2d62d15d40a73efae7f9a0bf0/src/proto/h1/decode.rs#L160-L164) to be hit:
  
  ```rs
  } else if num == 0 {
      return Poll::Ready(Err(io::Error::new(
          io::ErrorKind::UnexpectedEof,
          IncompleteBody,
      )));
  ```
  
  This branch is only hit when `num == 0` (bytes read from current chunk) _and_ `remaining > 0` (bytes remaining). Not sure if this is due to a hyper bug or a Deno bug. Commenting out that branch in `hyper` causes the body to be read successfully and a final chunk of length `0` to be read (which Deno treats as EoF and omits when converting to a stream), but that's presumably the wrong fix, as those numbers shouldn't be out of sync in the first place.
  
* `ec.europa.eu` uses `Transfer-Encoding: chunked`, with no `content-length` header, which is correct per the spec when using chunked transfer encoding. Commenting out the offending branch above does nothing here, and the hyper error's `cause.error` is "peer closed connection without sending TLS close_notify: https://docs.rs/rustls/latest/rustls/manual/_03_howto/index.html#unexpected-eof", which suggests it could be related to https://github.com/denoland/deno/issues/23960.

I don't think the `astroarts.co.jp` case can have anything to do with TLS, as the same thing occurs with the `http:` version. It's not possible to test `ec.europa.eu` sans TLS, as it just redirects to the `https:` version.

In both cases, capturing the TCP output in a file and then spinning up a TCP server with `Deno.listen` to respond with the contents of that file doesn't reproduce the error, so I guess both errors originate in the TCP layer. In the dumped `astroarts.co.jp` file, manually removing the `content-length` header does nothing; decrementing it truncates the result; and incrementing it errors on the same code path as the live version.

**Comment by lionel-rowe** [comment]
Further investigation: if `gzip` encoding is requested (which I guess Deno/its dependencies do internally by default), the `astroarts.co.jp` endpoint sets `content-encoding` to `gzip` and `content-length` to its _uncompressed_ length. This is wrong behavior, but Chrome and Firefox both accept it. Node and Bun both throw.

With `content-encoding: gzip`, behavior looks to be as follows (HTTP/1.1):

`content-length` header | Deno | Chrome | Firefox | Node | Bun | cURL
-|-|-|-|-|-|-
None (omitted) | OK | OK | OK | OK | OK | OK
Compressed length | OK | OK | OK | OK | OK | OK
Uncompressed length | Throws | OK | OK | Throws | Throws | OK
Arbitrarily shorter than compressed length | Throws | Truncates | Truncates | Truncates | Throws | Truncates
Arbitrarily longer than uncompressed length | Throws | Throws | OK | Throws | Throws | OK

The fact that Chrome makes a special allowance for cases where `content-length` is not the compressed length but _is_ the uncompressed length suggests this type of server misconfiguration may be common enough to warrant special-casing. None of the other clients special-case it, but Firefox and cURL both accept it (along with any other over-long value).

Meanwhile, manually setting `accept-encoding: identity` gives correct results in Deno, at the expense of the server sending more bytes than necessary over the wire:

```ts
const res = await fetch('https://www.astroarts.co.jp/article/feed.atom', {
    headers: {
        'accept-encoding': 'identity',
    },
})

console.log(await res.text())
```

The same fix also works for the `ec.europa.eu` URL... not sure if this indicates the root cause is related after all. `curl`ing it with `accept-encoding: gzip` gives neither a `content-length` header nor a `transfer-encoding`, while setting `accept-encoding: identity` gives `transfer-encoding: chunked`. Maybe the fact it switches to chunked transfer-encoding overrides the issue with the TCP connection as the final chunk `0\r\n` is more of a definitive indication of EOF than a zero-length message?

**Comment by marvinhagemeister** [comment]
cc @seanmonstar Maybe you know more about what's going on in hyper.

**Comment by seanmonstar** [comment]
The body error that hyper returns usually always includes more details in a source error. Would it be possible to use an error reporter when printing the error?

Regarding the length being the uncompressed length, yea that's clearly wrong. Some clients support it more by accident than by design, by just enforcing the body to be length advertised. hyper enforces the content-length to be exactly the number of bytes for a body. Being lenient here is a very dangerous idea, as that is what opens up possibilities for request smuggling.

**Comment by mahdiyari** [comment]
```
error: Uncaught (in promise) TypeError: error reading a body from connection
```
I couldn't figure out where this error happens in my code. There is no trace. I monkey patched fetch to catch its errors but this doesn't throw there.

I use ethers.js and that doesn't seem to use fetch. So I'm going to say this is not exclusive to fetch in Deno. This has been frustrating for me because it happens randomly and I don't have any idea where it happens exactly. Every single promise in my code is await inside try/catch. Non await promises have .catch().

I tried couple of things but couldn't figure out what call exactly throws this error.

**Comment by akku1139** [comment]
related? https://github.com/rustls/hyper-rustls/issues/162

**Comment by bartlomieju** [comment]
Thanks for the report. I dug into this, and it's a malformed response from the server, not a bug in Deno.

The server (`nginx/1.30.2`) sets `Content-Length` to the **uncompressed** size but then sends a gzip-encoded body that's much shorter, and closes the keep-alive connection early. Reproduced and traced with curl:

- No compression: server sends the full **16775** bytes, `Content-Length: 16775` — works fine.
- With `Accept-Encoding` (which `fetch` sends by default): response is `Content-Encoding: gzip` but **still** `Content-Length: 16775`, and the connection closes after only ~4838 bytes:
  ```
  curl: (18) transfer closed with 11937 bytes remaining to read   # 16775 - 4838
  ```
- The 4838-byte gzip stream itself is complete and valid (`gzip -t` passes; it decompresses to the full document).

So the server declares 16775 body bytes but only puts ~4838 on the wire. That violates HTTP — `Content-Length` must be the length of the encoded body actually transferred. Deno (via hyper) correctly rejects the truncated message with `error reading a body from connection`.

`curl` and browsers happen to tolerate this because the gzip stream is self-terminating, so they decode it and ignore the short transfer (curl still reports `error (18)` internally). Deno enforces the framing strictly, which is the correct behavior for a malformed response.

The fix belongs on the server side (it should send `Content-Length` equal to the compressed body length, or use chunked transfer encoding). Any change to make Deno lenient here — accepting a response when the content-decoding stream is complete even though the transport closed before `Content-Length` — would have to come from upstream hyper/reqwest, not Deno.

Closing as not a Deno bug. If you control the server or can reach the operator, the `Content-Length`/`Content-Encoding` mismatch is the thing to fix.
